walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,129 @@
+   * @return openTimeRanges
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<TimeRange> openTimeRanges = new ArrayList<>();
+
+    // When a the duration of the meeting is unrealistic and exceeds a day.
+    if(request.getDuration() > TimeRange.END_OF_DAY){
+      return openTimeRanges;
+    }
+
+    // When a meeting can take place any time on that day.
+    if (events.isEmpty()){
+      openTimeRanges.add(TimeRange.WHOLE_DAY);
+      return openTimeRanges;
+    }
+
+    /*****************
+     * General Case
+     */
+    Map<TimeRange, Set> attendeesMap = getAttendeesMap(events);
+    TreeSet<TimeRange> bookedTimeRanges = getSortedTimeRanges(events);
+
+    // Declare some helper variables
+    int start = TimeRange.START_OF_DAY;
+    TimeRange currentTimeRange = null;
+    long minimumDuration = 0;
+    Iterator<TimeRange> it = bookedTimeRanges.iterator();
+
+    // Iterate through the sorted TreeSet of event.
+    while(it.hasNext()){
+      currentTimeRange = it.next();
+      minimumDuration = start + request.getDuration();
+
+      // If all attendees are available, check if there is enough room for the requested event to take place.
+      // An if there is room, add TimeRange from start (i.e end of the previous event) to the beginning of the current event.
+      if(allAttendeesAreAvailable(new ArrayList<>(request.getAttendees()), new ArrayList<>(attendeesMap.get(currentTimeRange)))) {
+        if (minimumDuration <= currentTimeRange.start()) {
+          openTimeRanges.add(TimeRange.fromStartEnd(start, currentTimeRange.start(), false));
+        }
+        // Update start
+        start = currentTimeRange.end();
+      }
+    }
+
+    // Check if there enough time left of the day for the requested event
+    // If there is, add the the remaining time of the day
+    minimumDuration = start + request.getDuration();
+    if(minimumDuration <= TimeRange.END_OF_DAY){
+      openTimeRanges.add(TimeRange.fromStartEnd(start, TimeRange.END_OF_DAY, true));
+    }
+
+    return openTimeRanges;
+  }
+
+  /**********************************************************************************
+   * This method takes to list of attendees
+   * If at least one attendees appears in both list, it returns FALSE otherwise TRUE
+   * @param requestingAttendees
+   * @param bookedAttendees
+   * @return True/False
+   */
+
+  private boolean allAttendeesAreAvailable(ArrayList<String> requestingAttendees, ArrayList<String> bookedAttendees){
+      return requestingAttendees.removeAll(bookedAttendees);
+  }
+
+  /***********************************************************
+   * This method takes a collection of events and maps each
+   * TimeRange to its corresponding attendees.
+   *
+   * @param events
+   * @return attendeesMap
+   */
+
+  private Map<TimeRange, Set> getAttendeesMap(Collection<Event> events){
```

Comment:
Here the return type may better be "SortedMap" so that the caller of this API knows that it's not an ordinary map.